crates/iroha/tests/faulty_peers.rs

```diff
@@ -0,0 +1,519 @@
+
+                writer.write_all(&buf[..n]).await?;
+            }
+
+            Ok(())
+        }
+    }
+}
+
+async fn start_network_with_relay(network: &Network) -> Result<P2pRelay> {
+    let relay = P2pRelay::new(&network.peers().iter().map(|peer| peer.id()).collect());
+
+    timeout(
+        network.peer_startup_timeout(),
+        network
+            .peers()
+            .iter()
+            .enumerate()
+            .map(|(i, peer)| {
+                let topology = relay.topology_for(&peer.id());
+                let config = network
+                    .config()
+                    .write(["sumeragi", "trusted_peers"], &topology);
+                let genesis = genesis_factory(network.genesis_isi().clone(), topology);
+                async move {
+                    // FIXME: parallel
+                    peer.start(config, (i == 0).then_some(&genesis)).await;
+                    peer.once(|e| matches!(e, PeerLifecycleEvent::ServerStarted))
+                        .await;
+                }
+            })
+            .collect::<FuturesUnordered<_>>()
+            .collect::<Vec<_>>(),
+    )
+    .await?;
+
+    Ok(relay)
+}
+
+#[tokio::test]
+async fn network_starts_with_relay() -> Result<()> {
+    let network = NetworkBuilder::new().with_peers(4).build();
+    let mut relay = start_network_with_relay(&network).await?;
+
+    relay.start();
+    network.ensure_blocks(1).await?;
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn network_doesnt_start_without_relay() -> Result<()> {
+    let network = NetworkBuilder::new().with_peers(4).build();
+    let _relay = start_network_with_relay(&network).await?;
+
+    if let Ok(_) = timeout(
+        Duration::from_secs(3),
+        once_blocks_sync(network.peers().iter(), 1),
+    )
+    .await
+    {
+        panic!("network must not start!")
+    };
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn suspending_works() -> Result<()> {
+    const SYNC: Duration = Duration::from_secs(3);
+    const N_PEERS: usize = 4;
+    const { assert!(N_PEERS > 0) };
+
+    let network = NetworkBuilder::new().with_peers(N_PEERS).build();
+    let mut relay = start_network_with_relay(&network).await?;
+    // we will plug/unplug the last peer who doesn't have the genesis
+    let last_peer = network
+        .peers()
+        .last()
+        .expect("there are more than 0 of them");
+    let suspend = relay.suspend(&last_peer.id());
+
+    suspend.activate();
+    relay.start();
+
+    // all peers except the last one should get the genesis
+    timeout(
+        SYNC,
+        once_blocks_sync(network.peers().iter().take(N_PEERS - 1), 1),
+    )
+    .await?;
+    let Err(_) = timeout(SYNC, last_peer.once_block(1)).await else {
+        panic!("should not get block within timeout!")
+    };
+
+    // unsuspend, the last peer should get the block too
+    suspend.deactivate();
+    timeout(SYNC, last_peer.once_block(1)).await?;
+
+    Ok(())
+}
+
+// ======= ACTUAL TESTS BEGIN HERE =======
```

Comment:
Tests before this line are working (e.g. the relay is functioning). Tests below don't work yet for some weird reason, investigating.